pkg/controllers/node/node_controller.go

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package node
+
+import (
+	"reflect"
+	"sync"
+
+	"time"
+
+	calicocache "github.com/projectcalico/kube-controllers/pkg/cache"
+	"github.com/projectcalico/kube-controllers/pkg/controllers/controller"
+	"github.com/projectcalico/libcalico-go/lib/api"
+	"github.com/projectcalico/libcalico-go/lib/client"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/fields"
+	uruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	corecache "k8s.io/client-go/tools/cache"
+
+	"fmt"
+)
+
+// This cache maps a calicoNodeName to its kubernetesNodeName.
+type cache struct {
+	sync.RWMutex
+	nodes map[string]nodeData
+}
+
+type nodeData struct {
+	Name string
+}
+
+// NodeController implements the Controller interface.  It is responsible for monitoring
+// kubernetes nodes and responding to delete events by removing them from the Calico datastore.
+// It keeps a cache of known calico nodes and their corresponding kubernetes nodes to
+// accomplish this.
+type NodeController struct {
+	informer                corecache.Controller
+	k8sResourceCache calicocache.ResourceCache
+	nodeLookupCache   *cache
+	calicoClient            *client.Client
+	k8sClientset            *kubernetes.Clientset
+}
+
+
+// NewNodeController Constructor for NodeController
+func NewNodeController(k8sClientset *kubernetes.Clientset, calicoClient *client.Client) controller.Controller {
+	cacheArgs := calicocache.ResourceCacheArgs{
+		ObjectType: reflect.TypeOf(""),
+		ListFunc:   func() (map[string]interface{}, error) {
+			// Get all nodes from the Calico datastore
+			calicoNodes, err := calicoClient.Nodes().List(api.NodeMetadata{})
+			if err != nil {
+				return nil, err
+			}
+
+			// Iterate through and store the k8s nodes in our cache.
+			m := make(map[string]interface{})
+			for _, calicoNode := range calicoNodes.Items {
+				// find its kubernetes orchRef
+				k8sNodeName := getK8sNodeName(calicoNode)
```

Comment:
We only want to add the node to the map if it has a corresponding k8sNodeName.

So, I think we need to wrap this in an `if k8sNodeName != ""`